dbms/src/Dictionaries/CacheDictionary.cpp

```diff
@@ -610,10 +694,283 @@ void registerDictionaryCache(DictionaryFactory & factory)
+        const size_t max_update_queue_size =
+                config.getUInt64(layout_prefix + ".cache.max_update_queue_size", 100000);
+        if (max_update_queue_size == 0)
+            throw Exception{name + ": dictionary of layout 'cache' cannot have empty update queue of size 0",
+                            ErrorCodes::TOO_SMALL_BUFFER_SIZE};
+
+        const bool allow_read_expired_keys =
+                config.getBool(layout_prefix + ".cache.allow_read_expired_keys", false);
+
+        const size_t update_queue_push_timeout_milliseconds =
+                config.getUInt64(layout_prefix + ".cache.update_queue_push_timeout_milliseconds", 10);
+        if (update_queue_push_timeout_milliseconds < 10)
+            throw Exception{name + ": dictionary of layout 'cache' have too little update_queue_push_timeout",
+                            ErrorCodes::BAD_ARGUMENTS};
+
+        const size_t each_update_finish_timeout_seconds =
+                config.getUInt64(layout_prefix + ".each_update_finish_timeout_seconds", 600);
+        if (each_update_finish_timeout_seconds == 0)
+            throw Exception{name + ": dictionary of layout 'cache' cannot have timeout equals to zero.",
+                            ErrorCodes::BAD_ARGUMENTS};
+
+        const size_t max_threads_for_updates =
+                config.getUInt64(layout_prefix + ".max_threads_for_updates", 4);
+        if (max_threads_for_updates == 0)
+            throw Exception{name + ": dictionary of layout 'cache' cannot have zero threads for updates.",
+                            ErrorCodes::BAD_ARGUMENTS};
+
+        return std::make_unique<CacheDictionary>(
+                database, name, dict_struct, std::move(source_ptr), dict_lifetime, size,
+                allow_read_expired_keys, max_update_queue_size, update_queue_push_timeout_milliseconds,
+                each_update_finish_timeout_seconds, max_threads_for_updates);
     };
     factory.registerLayout("cache", create_layout, false);
 }
 
+void CacheDictionary::updateThreadFunction()
+{
+    setThreadName("AsyncUpdater");
+    while (!finished)
+    {
+        UpdateUnitPtr first_popped;
+        update_queue.pop(first_popped);
+
+        if (finished)
+            break;
+
+        /// Here we pop as many unit pointers from update queue as we can.
+        /// We fix current size to avoid livelock (or too long waiting),
+        /// when this thread pops from the queue and other threads push to the queue.
+        const size_t current_queue_size = update_queue.size();
+
+        if (current_queue_size > 0)
+            LOG_TRACE(log, "Performing bunch of keys update in cache dictionary with "
+                            << current_queue_size + 1 << " keys");
+
+        std::vector<UpdateUnitPtr> update_request;
+        update_request.reserve(current_queue_size + 1);
+        update_request.emplace_back(first_popped);
+
+        UpdateUnitPtr current_unit_ptr;
+
+        while (update_queue.tryPop(current_unit_ptr))
```

Comment:
The comment above says that we don't try to pop the entire queue, only the number of items we've seen first time. But this loop looks like it actually pops everything.

Reply:
It's true. But I don't know how to do it better (code will be too complex because of additional counter etc.)